src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/env.hpp>
+#include <miopen/errors.hpp>
+#include <miopen/logger.hpp>
+#include <miopen/stringutils.hpp>
+#include <amd_comgr.h>
+#include <algorithm>
+#include <exception>
+#include <cstddef>
+#include <cstring>
+#include <tuple> // std::ignore
+#include <vector>
+
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_COMGR_LOG_CALLS)
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_COMGR_LOG_OPTIONS)
+/// Integer, set to max number of first characters
+/// you would like to log onto console.
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_COMGR_LOG_SOURCE_TEXT)
+
+/// \todo see issue #1222, PR #1316
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_SRAM_EDC_DISABLED)
+
+#define COMPILER_LC 1
+
+#define EC_BASE(comgrcall, info, expr2)                                   \
```

Comment:
I think these macros should be replaced by function calls because it makes the code harder to follow. Its not clear what gets evaluated when. To make it so we can get the function name, we can lift the function into an object and use `__PRETTY_FUNCTION__` to get the function name:

```cpp

template <class MIOpen_Private_FunctionType_, MIOpen_Private_FunctionType_ MIOpen_Private_Function_>
const std::string& get_function_name()
{
    static std::string name;

    if(name.empty())
    {
#ifdef _MSC_VER
        name = typeid(MIOpen_Private_FunctionType_).name();
        name = name.substr(7);
#else
        const char parameter_name[] = "MIOpen_Private_Function_ =";

        name = __PRETTY_FUNCTION__;

        auto begin  = name.find(parameter_name) + sizeof(parameter_name);
#if(defined(__GNUC__) && !defined(__clang__) && __GNUC__ == 4 && __GNUC_MINOR__ < 7)
        auto length = name.find_last_of(",") - begin;
#else
        auto length = name.find_first_of("];", begin) - begin;
#endif
        name        = name.substr(begin+1, length-1);
#endif
    }

    return name;
}

template<class F, F f>
struct function
{
    template<class... Ts>
    decltype(auto) operator()(Ts&&... xs) const
    {
        return f(std::forward<Ts>(xs)...);
    }

    const std::string& name() const
    {
        return get_function_name<F, f>();
    }
};

// So we dont have to repeat the function name twice
#define MIOPEN_LIFT(...) function<decltype(&__VA_ARGS__), __VA_ARGS__>
```

And then `ec_base` could be implemented like this:

```cpp
template<class ComgrCall, class Info, class Action>
auto ec_base(ComgrCall cc, Info info, Action action)
{
    return [=](auto... xs) {
        const amd_comgr_status_t status = cc(xs...);
        if(status != AMD_COMGR_STATUS_SUCCESS)
        {
            MIOPEN_LOG_E("\'" << cc.name() << "\' " << to_string(info) << ": " << GetStatusText(status));
            action();
        }
        else if(miopen::IsEnabled(MIOPEN_DEBUG_COMGR_LOG_CALLS{}))
        {
            MIOPEN_LOG_E("\'" << cc.name() << "\' " << to_string(info));

        }
    }
}
```

And the rest of the macros can be implemented as function on top of that. Then when you want to call a function just do:

```cpp
template<class ComgrCall>
auto ec(ComgrCall cc)
{
    return ec_base(cc, NoInfo, []{});
}

ec(MIOPEN_LIFT(amd_comgr_release_data))(handle);
```

There is still a macro but it is only used to lift the function(so we dont have to write it twice), the rest is all normal C++ code.

Reply:
Why not just use type erasure in function<...> to get rid of templates in uses (like ec(...)) to have clearer types and faster compilation?

Reply:
Because not all the function have the same signature and the function pointer has to be passed as a template parameter to be able to deduce the function name.

Reply:
@pfultz2 Thanks, I think that this is really great idea and appreciate your efforts, however I would better implement it separately, when it becomes easier for me to follow this new-style code and I have some free time for refactoring.